### PR TITLE
feat: Add Bert embedding torch_batch_process example

### PR DESCRIPTION
**Pull request by szewaiyuen6**
## Description

Add torch_batch_process example which uses Bert to generate embedding and upload to Chroma vector db

## Test Plan

coverage run -m pytest -vv --capture=tee-sys --disable-warnings --master-host="<det_cluster_addr>" --master-port="8080" --durations=0 tests/nightly/test_distributed.py::test_torch_batch_process_generate_embedding

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: examples/features/torch_batch_process/batch_inference/generate_embedding/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+accelerate>=0.12.0
+transformers>=4.28.1,<4.29.0
```
Comment (rb-determined-ai) [inline examples/features/torch_batch_process/batch_inference/generate_embedding/requirements.txt:2]: what is the `<4.29.0` limit for?

**Review by rb-determined-ai — APPROVED** [review]
this seems fine to me but also I don't really know anything about embeddings so prolly get an amle to look over it too.

File: examples/features/torch_batch_process/batch_inference/generate_embedding/bert_embedding_generation.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            masks = masks.to(self.device)
+
+            outputs = self.model(inputs, masks)
+            outputs = torch.mean(outputs["hidden_states"][-1], dim=1)
+
+            self.output.append({"embeddings": outputs, "id": batch["_id"], "text": batch["text"]})
+            self.last_index = batch_idx
+
+    def on_checkpoint_start(self):
+        if len(self.output) == 0:
```
Comment (aciborowska) [inline examples/features/torch_batch_process/batch_inference/generate_embedding/bert_embedding_generation.py:52]: nit: can we add some explanation in `on_checkpoint_start` and `on_finish` to explain to the user what is happening in the code? I think that a novice user that wants to create a vector db, may not know immediately e.g., (1) why file name should contain rank or last index; (2) if the data is saved to a checkpoint; (3) that `on_finish` is where vector db is created after all documents have been processed.

File: examples/features/torch_batch_process/batch_inference/generate_embedding/bert_embedding_generation.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                max_length=512,
+                add_special_tokens=True,
+            )
+            inputs = torch.tensor(tokenized_text["input_ids"])
+            inputs = inputs.to(self.device)
+            masks = torch.tensor(tokenized_text["attention_mask"])
+            masks = masks.to(self.device)
+
+            outputs = self.model(inputs, masks)
+            outputs = torch.mean(outputs["hidden_states"][-1], dim=1)
```
Comment (aciborowska) [inline examples/features/torch_batch_process/batch_inference/generate_embedding/bert_embedding_generation.py:46]: I'd add here some text explaining how embeddings are built since we are using one of many simple strategies. For instance: To create an embedding vector for each document, first, we take the hidden states from the last layer (output["hidden_states"][-1]), which results in a tensor of (#examples, #tokens, #hidden_states) size. Next, we calculate the average across the token-dimension, resulting in (#examples, #hidden_states) vector.

File: examples/features/torch_batch_process/batch_inference/generate_embedding/README.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+## Use case
+
+One common LLM use case is information retrieval. The first step of information retrieval is to generate embedding on
+documents and upload to a vector database. When a user query come in, the system would retrieve documents most relevant
+to the query from the vector database. These documents are often used to supplement the prompt provided to the LLM to 
+improve the quality of answer.
+
+## How does torch_batch_process API help?
+
+In this example, we use the `torch_batch_process` API to
```
Comment (aciborowska) [inline examples/features/torch_batch_process/batch_inference/generate_embedding/README.md:12]: nits: 
* `the system would retrieve` -> `the system would first compute embedding of the query, and then retrieve...`
* `API to` -> `API to:`
* `4 workers` -> `4 workers,`

**Review by aciborowska — APPROVED** [review]
a few nits, but look good to me otherwise!

File: examples/features/torch_batch_process/batch_inference/generate_embedding/requirements.txt (hunk 1 above)
Comment (szewaiyuen6) [reply examples/features/torch_batch_process/batch_inference/generate_embedding/requirements.txt:2]: I put it in to align with other [determined requirements.txt](https://github.com/search?q=repo%3Adetermined-ai%2Fdetermined%20%3C4.29.0&type=code) files that specify it.
